src/_tkagg.cpp

```diff
@@ -95,8 +96,101 @@ static PyObject *mpl_tk_blit(PyObject *self, PyObject *args)
     }
 }
 
+#ifdef WIN32_DLL
+LRESULT CALLBACK
+DpiSubclassProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam,
+                UINT_PTR uIdSubclass, DWORD_PTR dwRefData)
+{
+    switch (uMsg) {
+    case WM_DPICHANGED:
+        // This function is a subclassed window procedure, and so is run during
```

Comment:
Consider using `Tcl_QueueEvent` or `Tcl_EvalEx` to queue up a call to a tkinter-registered python function to avoid the polling loop. Failing that, I think this all occurs in the main thread, so you can use a `deque` instead of a `SimpleQueue`.

Reply:
It is not possible to call any Tk functions due to the Tkinter-internal Tcl thread lock. I'm fairly certain I tried `Tcl_QueueEvent` as well, but can do so again.

Reply:
I was under the impression that these windowprocs are called within `Tcl_DoOneEvent`, usually here: 

https://github.com/python/cpython/blob/1e9f0933095403b215c2c4a0be7915d034ff7026/Modules/_tkinter.c#L2938

so I was expecting it to hold the `tcl_lock` already. If that's not the case... sorry for the distraction!

Reply:
Actually no, I think I tried `Tk_QueueEvent` which does not work because of said locks. Looking at the API for `Tcl_QueueEvent`, I'm not sure how it could be used. It takes a pointer a function to call later, which would still be my code and unable to (un)lock anything in the Tkinter library.

Reply:
Note that Tk/Tcl is threaded on Windows, so it doesn't go through that route, but this one:
https://github.com/python/cpython/blob/1e9f0933095403b215c2c4a0be7915d034ff7026/Modules/_tkinter.c#L2931

As an alternative, I tried calling `Tcl_SetVar` on a variable named per-window, and this might work without crashing. I need to do a bit more testing to be sure, since concurrency issues can be heisenbugs.

Reply:
OK, I used a Tk variable now; this is nice as it cleans up the queue polling.

I think what happened originally is that I tried the variable approach from Python, and that failed due to the competing locks which I only found out about some time later.

Reply:
I like it, way better than the original and less magical than what I was imagining.